### ISSUE TITLE
Janus dead lock when using sip plugin

### ISSUE THREAD

**Issue by chraja1233** [meetecho/janus-gateway#798]
Hi,
I'm using Janus SIP plugin for calls, and have found this issue for several times in load test.

**ISSUE:**
Connected client returns LWS_CALLBACK_CLIENT_CONNECTION_ERROR due to TIMEDOUT WAITING on 4 reason and again try to connect the WebSocket won't accept any new client after some register requests

After tracing the code execution, I found the code block at janus_websockets_send_message() function in transports/janus_websockets.c

       janus_mutex_lock(&client->mutex);
        /* Convert to string and enqueue */
        char *payload = json_dumps(message, json_format);
        g_async_queue_push(client->messages, payload);
#ifdef HAVE_LIBWEBSOCKETS_NEWAPI
        printf("RAJA janus_websockets_send_message1 lws_callback_on_writable\n");
        lws_callback_on_writable(client->wsi);
        printf("RAJA janus_websockets_send_message2 lws_callback_on_writable\n");
#else
        libwebsocket_callback_on_writable(client->context, client->wsi);
#endif

Mutex is locked then call lws_callback_on_writable but it is never return so lock is never unlocked.

I think this issue might be related to 
https://github.com/meetecho/janus-gateway/issues/732

Thanks & Regards,
Raja Chegondi.

**Comment by lminiero** [comment]
Thanks for the heads up, I'll look into it. Not sure how we can have a mutex lock without an unlock and it only deadlocks from time to time, though?

**Comment by lminiero** [comment]
The unlock is there:

https://github.com/meetecho/janus-gateway/blob/master/transports/janus_websockets.c#L936

are you sure you didn't remove it when adding your text? We do lock the mutex when the `LWS_CALLBACK_SERVER_WRITEABLE` event is called:

https://github.com/meetecho/janus-gateway/blob/master/transports/janus_websockets.c#L1203 

but that happens within the libwebsockets thread, and not the core thread that invokes `janus_websockets_send_message`, and so they should never deadlock.

**Comment by chraja1233** [comment]
Thanks for your quick reply Lorenzo.
Please find our observation below
Mutex lock was placed just before invoking lws_callback_on_writable API and mutex unlock was just after this API. Now what we have observed is, control was never returning back from lws_callback_on_writable API, hence the dead lock arises.

**Comment by lminiero** [comment]
Got it: as I said, `lws_callback_on_writable` should not be blocking, as it should simply trigger something in the libwebsockets event loop that should result in a `LWS_CALLBACK_SERVER_WRITEABLE` event. This event, though, should happen within the libwebsockets thread, as it's triggered manually by a call to `lws_service`:

https://github.com/meetecho/janus-gateway/blob/master/transports/janus_websockets.c#L996

and not automatically as a result of `lws_callback_on_writable`. So I really don't believe it's one of our locks that deadlocks, but something else, possibly a libwebsockets internal? Anyway, you can check that for sure, by adding some prints before and after that lock and try multiple connections to see if it's our lock that blocks things in your case or not.

**Comment by lminiero** [comment]
Any update? Got any new printouts to share?

**Comment by chraja1233** [comment]
Dead lock also occurred in video room plugin also. There is no link with load. Right now, i am debugging where it is blocking. Once i get any information i will update.

**Comment by lminiero** [comment]
Closing as most likely a duplicate of #732 (SIP has nothing to do with it).
